File: binary-tree-right-side-view/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None: # No tree exists
            return None

        answer = [] # Answer to return
        current_level = 1 # Count nodes on current level
        next_level = 0 # Count nodes on next level
        queue = [root]
        while queue: # BFS
            node = queue.pop(0)
            if node.left:
                queue.append(node.left)
                next_level += 1
            if node.right:
                queue.append(node.right)
                next_level += 1

            current_level -= 1
            if current_level == 0: # If end of current level, remember rightmost node
                answer.append(node.val)
                current_level = next_level
                next_level = 0
        return answer

File: binary-tree-right-side-view/binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    public List<Integer> rightSideView(TreeNode root) 
    {
        // No tree exists, return empty list
        List<Integer> answer = new ArrayList<>();
        if (root == null) 
            return answer;
        
        // Setup for BFS and track nodes on each level
        Deque<TreeNode> queue = new LinkedList<>();
        queue.addLast(root);
        int currentLevel = 1, nextLevel = 0;
        
        // BFS
        while (!queue.isEmpty())
        {
            // Pop the node
            TreeNode node = queue.removeFirst();
            currentLevel -= 1;
            
            // Check left
            if (node.left != null) 
            {
                queue.addLast(node.left);
                nextLevel += 1;
            }
            // Check right
            if (node.right != null)
            {
                queue.addLast(node.right);
                nextLevel += 1;
            }
            
            // End of current level, add to right-side list
            // Update level counts
            if (currentLevel == 0)
            {
                answer.add(node.val);
                currentLevel = nextLevel;
                nextLevel = 0;
            }
        }
        
        return answer;
    }
}